pkg/auth/dex.go

```diff
@@ -0,0 +1,247 @@
+		return
+	}
+
+	// TODO(BB) Set an app state to make the connection more secure
+	authCodeURL := oauthConfig.AuthCodeURL("APP_STATE")
+	http.Redirect(w, r, authCodeURL, http.StatusSeeOther)
+}
+
+// HandleCallback is the callback handler for an OAuth2 login flow.
+func (a *DexAppClient) handleCallback(w http.ResponseWriter, r *http.Request) {
+	oauth2Config, err := a.oauth2Config(nil)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if errMsg := r.FormValue("error"); errMsg != "" {
+		errorDesc := r.FormValue("error_description")
+		http.Error(w, html.EscapeString(errMsg)+": "+html.EscapeString(errorDesc), http.StatusBadRequest)
+		return
+	}
+
+	code := r.FormValue("code")
+	ctx := oidc.ClientContext(r.Context(), a.Client)
+	token, err := oauth2Config.Exchange(ctx, code)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get token: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	idTokenRAW, ok := token.Extra("id_token").(string)
+	if !ok {
+		http.Error(w, "no id_token in token response", http.StatusInternalServerError)
+		return
+	}
+
+	idToken, err := a.validateToken(ctx, idTokenRAW, a.ClientID)
+	if err != nil {
+		http.Error(w, "failed to verify the token", http.StatusInternalServerError)
+		return
+	}
+
+	var claims jwt.MapClaims
+	err = idToken.Claims(&claims)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Stores the oauth token into the cookie.
+	if idTokenRAW != "" {
+		expiration := time.Now().Add(time.Duration(expirationDays) * 24 * time.Hour)
+		cookie := http.Cookie{
+			Name:    dexOAUTHTokenName,
+			Value:   idTokenRAW,
+			Expires: expiration,
+			Path:    "/",
+		}
+		http.SetCookie(w, &cookie)
+	}
+	http.Redirect(w, r, a.BaseURL, http.StatusSeeOther)
+}
+
+func (a *DexAppClient) validateToken(ctx context.Context, rawToken string, allowedAudience string) (token *oidc.IDToken, err error) {
+	p, err := oidc.NewProvider(ctx, a.IssuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	// Token must be verified against an allowed audience.
+	config := oidc.Config{ClientID: allowedAudience}
+	verifier := p.Verifier(&config)
+	idToken, err := verifier.Verify(ctx, rawToken)
+	if err != nil {
+		return nil, fmt.Errorf("the token could not be verified, audience %s is not allowed with err: %s", allowedAudience, err)
+	}
+
+	return idToken, nil
+}
+
+func (a *DexAppClient) oauth2Config(scopes []string) (c *oauth2.Config, err error) {
+	ctx := oidc.ClientContext(context.Background(), a.Client)
+	p, err := oidc.NewProvider(ctx, a.IssuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &oauth2.Config{
+		ClientID:     a.ClientID,
+		ClientSecret: a.ClientSecret,
+		Endpoint:     p.Endpoint(),
+		Scopes:       scopes,
+		RedirectURL:  a.RedirectURI,
+	}, nil
+}
+
+// TODO(BB): Verifies if the token has the cookie and validates it.
+func (a *DexAppClient) verifyToken(r *http.Request) error {
```

Comment:
Any reason why this function is not finished?

Reply:
I did not want to make the PR too extensive, but I guess one method wont make a big difference 😅 I added it.